Check for overflow during generation of VLAs

fbrausse opened the pull request:
This PR addresses #45 by implementing @jmorse's suggestions:
- change the type for the computation of the size from uint_type() to size_type(), and
- add an !overflow assertion for this computation.

I've added 4 test cases, 32 and 64 bit, #if 0 and 1. The --64 ones do not have the overflow (and are VERIFICATION UNKNOWN with the --max-k-step 5), while the --32 ones do both show the overflow in the counter-example.

fbrausse committed:
[goto-convert] check for overflow during generation of VLAs

Fixes #45.

fbrausse committed:
[regression] add cases for #45

lucasccordeiro reviewed (approved):
Thanks, @fbrausse.

Shall we also run this PR over the SV-COMP benchmarks?

fbrausse commented:
@lucasccordeiro Thanks! Regarding SV-COMP I don't know. It adds an assertion that is about overflows of `sizeof(vla)`. Can't really tell whether they have nondet-ly sized VLAs that big. Your choice. ;)

*Edit:* As the trigger via Github is working again, it's easy to do now.

fbrausse committed:
[regression] adjust verdict for VLA cases w/o upper bound on size

fbrausse committed:
[regression] add successful case with upper bound on size

fbrausse committed:
[goto-convert] gen-vla: check all arithmetic for overflows

fbrausse commented:
We have 2 regression test cases that fail with this PR:
- regression/esbmc/multi_dimensional_array_1
- regression/esbmc/multi_dimensional_array_2

Both assume positive constants with no upper bound and then create an `int arr[N][M]` or `int arr[N][M][O]`, respectively. This obviously doesn't fit, so I've adjusted their expected verdict. Instead of those original ones, there is a new test case that is successful, where we assume `N < SIZE_MAX / sizeof(int) / M` in order to avoid the overflow. This locally takes about 1 minute to solve, even without the loop the original had. It's marked THOROUGH. Let's see what the CI says about it.

I've also added assertions all along the way during the computation, because, unintuitively for me, the overflow* expressions only check their immediate operand, not all the way down to the leaves. Since it could be many more assertions now, I'll start an SV-COMP run to see the resulting fallout.

fbrausse commented:
I'm wondering whether this should be behind `--overflow-check` or not. Any suggestions?

lucasccordeiro commented:
> I'm wondering whether this should be behind --overflow-check or not. Any suggestions?

This option looks reasonable to me.

fbrausse commented:
[Master](https://github.com/esbmc/esbmc/actions/runs/5540520772) (18d016a377ffe9e73b1b4ef40c9cabd90a068d39):
```
Statistics:          26459 Files
  correct:           14415
    correct true:     8194
    correct false:    6221
  incorrect:            61
    incorrect true:     28
    incorrect false:    33
  unknown:           11983
  Score:             21185 (max: 43043)
```
[This PR](https://github.com/esbmc/esbmc/actions/runs/5559582024) (acd8ca80f4e1d884241928c85989814fc8b9a95c):
```
Statistics:          26459 Files
  correct:           14381
    correct true:     8168
    correct false:    6213
  incorrect:           115
    incorrect true:     28
    incorrect false:    87
  unknown:           11963
  Score:             20261 (max: 43043)
```
This is clearly not SV-COMP-compatible because many of the new incorrect benchmarks are just like the #45 example. We also cannot put it just behind the `--overflow-check` as this option is already used by the esbmc-wrapper.py for checking those benchmarks.

fbrausse commented:
@mikhailramalho Given that you authored #45, do you have a suggestion how to proceed here? Do you agree with the approach in this PR? If so, do you think we should opt-in or opt-out of checking these properties?

lucasccordeiro commented:
@fbrausse: we should raise an issue in SV-COMP about this problem. In the meantime, I would also recommend adding some options to enable or disable that kind of check so we can progress with our development.

What do you think?

mikhailramalho commented:
It looks correct to me, my only suggestion is to move the integer/real encoding to the smt backend; it should be already there anyway, we shouldn't check overflow in these cases.

I agree with Lucas, we should fix the tests instead.

fbrausse commented:
Thanks for your comments!

> we should raise an issue in SV-COMP about this problem.

Will do, but judging from previous experience this will take quite some time.

> In the meantime, I would also recommend adding some options to enable or disable that kind of check so we can progress with our development.

Right. Should it be opt-in or opt-out?

> It looks correct to me, my only suggestion is to move the integer/real encoding to the smt backend; it should be already there anyway, we shouldn't check overflow in these cases.

It is in the backend, yes. I've disabled generating the asserts with `--ir` so we don't need to process them everywhere else down the line when they'll get thrown away in smt anyway. I thought it'd make the intermediate steps like the slicer more efficient to have fewer useless asserts. Functionally there's no difference, I can remove the checks if you think that's better.

mikhailramalho commented:
Em seg., 17 de jul. de 2023 às 13:39, Franz Brauße ***@***.***>
escreveu:

> Thanks for your comments!
>
> we should raise an issue in SV-COMP about this problem.
>
> Will do, but judging from previous experience this will take quite some
> time.
>

Indeed, people will probably only start to look at issues again when it's
closer to the conference.

> In the meantime, I would also recommend adding some options to enable or
> disable that kind of check so we can progress with our development.
>
> Right. Should it be opt-in or opt-out?
>
> It looks correct to me, my only suggestion is to move the integer/real
> encoding to the smt backend; it should be already there anyway, we
> shouldn't check overflow in these cases.
>
> It is in the backend, yes. I've disabled generating the asserts with --ir
> so we don't need to process them everywhere else down the line when they'll
> get thrown away in smt anyway. I thought it'd make the intermediate steps
> like the slicer more efficient to have fewer useless asserts. Functionally
> there's no difference, I can remove the checks if you think that's better.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/1190#issuecomment-1638495124>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHZLPZIEWPNJ3FJ4X7LXQVTEFANCNFSM6AAAAAA2H67OEU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail R. Gadelha, Ph.D.

fbrausse committed:
[util] fix wrong type in type_byte_size_bits_expr() computation

Fixes #1198.

fbrausse committed:
[options] add --no-vla-size-check to disable checking VLA sizes for overflows

fbrausse committed:
[svcomp] let wrapper use the new --no-vla-size-check

This check is incompatible with current SV-COMP.

fbrausse committed:
[goto-convert] do not check "int-encoding"

fbrausse commented:
I've made the checks opt-out as the default verification results should be a bit more sane than what SV-COMP assumes. The flag is called `--no-vla-size-check`. The "int-encoding" check is also gone.

lucasccordeiro commented:
Thanks, @fbrausse: Can you please fix the conflicting file `.github/workflows/benchexec.yml`?

lucasccordeiro commented:
Thanks for submitting this PR, @fbrausse.
